pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,221 @@
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+//isCloudKubernetesServiceIndexer
+const (
+	isCKSIndexer = "clusters.management.cattle.io/is-cks"
+)
+
+var (
+	// for other cloud drivers, please edit HERE
+	cloudDrivers             = []string{apimgmtv3.ClusterDriverEKS}
+	clusterUpstreamRefresher *clusterRefreshController
+)
+
+// for other cloud drivers, please edit HERE
+type clusterConfig struct {
+	eksConfig *eksv1.EKSClusterConfigSpec
+}
+
+func init() {
+	// possible settings controller, which references refresh
+	// cron job, will run prior to .
+	// This ensure the CronJob will not be nil
+	clusterUpstreamRefresher = &clusterRefreshController{
+		refreshCronJob: cron.New(),
+	}
+}
+
+type clusterRefreshController struct {
+	refreshCronJob *cron.Cron
+	secretsCache   wranglerv1.SecretCache
+	clusterClient  v3.ClusterClient
+	clusterCache   v3.ClusterCache
+}
+
+func StartClusterUpstreamCronJob(wContext *wrangler.Context) {
+	clusterUpstreamRefresher.secretsCache = wContext.Core.Secret().Cache()
+	clusterUpstreamRefresher.clusterClient = wContext.Mgmt.Cluster()
+	clusterUpstreamRefresher.clusterCache = wContext.Mgmt.Cluster().Cache()
+
+	clusterUpstreamRefresher.clusterCache.AddIndexer(isCKSIndexer, func(obj *apimgmtv3.Cluster) ([]string, error) {
+		// for other cloud drivers, please edit HERE
+		switch {
+		case obj.Spec.EKSConfig != nil:
+			if obj.Status.EKSStatus.UpstreamSpec == nil {
+				logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", obj.Name)
+				return []string{}, nil
+			}
+			return []string{apimgmtv3.ClusterDriverEKS}, nil
+		default:
+			return []string{}, nil
+		}
+	})
+
+	schedule, err := cron.ParseStandard(settings.ClusterUpstreamRefreshCron.Get())
+	if err != nil {
+		logrus.Errorf("Error parsing upstream cluster refresh cron. Upstream state will not be refreshed: %v", err)
+		return
+	}
+	clusterUpstreamRefresher.refreshCronJob.Schedule(schedule, cron.FuncJob(clusterUpstreamRefresher.refreshAllUpstreamStates))
+	clusterUpstreamRefresher.refreshCronJob.Start()
+}
+
+func (c *clusterRefreshController) refreshAllUpstreamStates() {
+	logrus.Debugf("Refreshing clusters' upstream states")
+	for _, cloudDriver := range cloudDrivers {
+		clusters, err := c.clusterCache.GetByIndex(isCKSIndexer, cloudDriver)
+		if err != nil {
+			logrus.Error("error trying to refresh clusters' upstream states")
+			return
+		}
+
+		for _, cluster := range clusters {
+			if _, err := c.refreshClusterUpstreamSpec(cluster, cloudDriver); err != nil {
+				logrus.Errorf("error refreshing cluster [%s] upstream state", cluster.Name)
+			}
+		}
+	}
+}
+
+func (c *clusterRefreshController) refreshClusterUpstreamSpec(cluster *mgmtv3.Cluster, cloudDriver string) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
```

Comment:
Was looking for the error check here but looks like it moved above to the start function.

Reply:
Yes, it was moved to `func StartClusterUpstreamCronJob()`